Track if control-flow statements have any label inside in any nesting level.

baziotis opened the pull request:
*Why?*
To be able to query if control-flow statements have any label inside.

*Double Why?*
Because it helps LDC back-end do dead-code elimination when the condition is constant: [PR](https://github.com/ldc-developers/ldc/pull/3134).
But if the body to-be-elided has any label, we can't eliminate it because it may be accessed through the label. So, currently, we have to walk the tree again to get that info.

*How?*
If we find a label, we propagate it backwards through the scopes. Then, the statements that we care to keep that info just query the scope.
When we go forward the scopes again (i.e. push), the info is reset.

Note: This is just a draft implementation in order to see if you're positive on the addition. If yes, the prospect is to do what has been done to `if` for `while`, `for` etc. and also treat `CaseStatement` / `DefaultStatement` as label if they're not directly followed by a `SwitchStatement`.

baziotis committed:
Track if an  has any label inside

baziotis commented on src/dmd/statementsem.d:
```diff
@@ -2262,6 +2265,10 @@ else
 
         ifs.ifbody = ifs.ifbody.semanticNoScope(scd);
+        if (scd.has_label) {
+            ifs.ifbody.has_label = true;
```
This might actually not be better to set `ss.has_label` in `visit(ScopeStatement)` (and maybe test here on the `if.body`).

thewilsonator commented:
Please also update the header files with the added fields.

baziotis committed:
Update statement.h with has_label field

thewilsonator reviewed (approved):
Thanks, lmk when this is good.

JohanEngelen commented:
`case:` and `default:` are not handled while they should.
How will this be tested?

baziotis commented:
> `case:` and `default:` are not handled while they should.

Yes, because I did the bare minimum to see if people are interested.

> How will this be tested?

This is puzzling me as well. One has to access the AST to test this.

> Thanks, lmk when this is good.

Thanks for the approval. Should I do everything in this PR? Here's what has to be done (at minimum):
- `case` and `default`
- `for`, `while`, `do`. `foreach` might be a special case.
Maybe `switch`...

Before moving forward though, I think it's important to find a way to test.

baziotis committed:
Updated scope.h fields and added case/default handling

baziotis commented on src/dmd/dscope.d:
```diff
@@ -75,6 +75,10 @@ struct Scope
     ScopeDsymbol scopesym;          /// current symbol
     FuncDeclaration func;           /// function we are in
     Dsymbol parent;                 /// parent to use
+    bool has_label;                 /// There is a label somewhere inside (in any nesting level)
```
This may not be the best place to put a `bool` due to packing etc.

baziotis committed:
Commented printf

baziotis commented:
I added the fields to `scope.h` and added handling for `case/default`. The last 2 were not as easy as I had imagined. It can't be handled as I did in LDC (or at least it will be very inefficient).
_The LDC method_: In LDC, I had to check _separately_ for each `if` if it has a label.
That means that if we find an `if`, we visit its body (notice, we're going from the `if` and inside / forwards only).
If we find a `case` and we haven't seen a `switch` on the way, then it acts as a label.

Here it works backwards. You can see the PR description the simple case of labels. The problem
with `case`s is that since we're going from the `case` and backwards, we might see a lot of `if`s on the way (notice that we don't start from a specific `if` as in LDC). Basically, that means that for every `case`, we need to know its enclosing `switch` so we can test whether any `if` we come across is inside that or not.

_Testing_: As Johan had mentioned, it doesn't seem that there is a way to test. The best I could think of is to add a `printf` for every `if` that is marked as having a label and provide this "testcase".
This helped me and it may at least help the reviewers.

```
void foo(int x) {
    switch (x) {
    case 1:
        if (1) {  // Has label
    case 2: assert(0);
        }
        break;
    default: assert(0);
    }
}

void bar(int x) {
    if (2) {  // Doesn't have label
        switch (x) {
        case 1:
        case 2: assert(0);
            break;
        default: assert(0);
        }
    }
}

void three(int a, int b, int c) {
L1:
    if (5) {  // Has label
        if (3) {  // Has label
        L2:
            int d;
        }
        if (4) {  // Doesn't have label
            int d = 2;
        }
    }
}

void four(int a, int b, int c) {
L1:
    if (7) {  // Has label
        if (6) {  // Has label
        L2:
            int d;
        }
        
    }
    if (8) {  // Doesn't have label
        int d = 2;
    }
}
```

@thewilsonator I'm still thinking the method through but if I haven't missed something, this is finished. I'd like to see this (which handles only `if`s) getting merged to continue with the rest. In any case,
it contains most hard parts.

lesderid reviewed (commented):
`has_label` should be named `hasLabel` or `haslabel` for style consistency. Same for other variables.

rainers commented:
> _Testing_: As Johan had mentioned, it doesn't seem that there is a way to test.

Can you do it as a frontend unit test like https://github.com/dlang/dmd/blob/master/test/unit/frontend.d#L223 that can check the AST?

baziotis committed:
Naming convention change

baziotis commented:
> 
> Can you do it as a frontend unit test like https://github.com/dlang/dmd/blob/master/test/unit/frontend.d#L223 that can check the AST?

Thanks I didn't know about that. Maybe ? I'm not sure how to specifically target an `if`.

rainers commented:
> Thanks I didn't know about that. Maybe ? I'm not sure how to specifically target an `if`.

I guess that would either mean "search the members of the module for the function declaration, than search its body statement for the respective if-statement recursively" or "write a small visitor that handles IfStatements".

baziotis commented:
Yes, thank you. It seems it can be handled with a visitor like [this](https://github.com/dlang/dmd/blob/master/test/unit/frontend.d#L11). But currently I can't compile (actually link) with this branch's dmd so I'll see how can I fix this first.

baziotis committed:
Added unit tests

JohanEngelen commented:
`hasLabelBecauseOfLabel ` is a strange variable name and indicates the bigger problem: I think what you want semantic analysis to do here is not to mark when a block contains a "label". What you want to check is whether the block can be jumped into; whether there is control flow that can enter the block at another position than the block's start. I suggest changing the naming of a number of variables according to that. Something like `hasMultipleEntryPoints`.

baziotis commented:
> `hasLabelBecauseOfLabel ` is a strange variable name and indicates the bigger problem: I think what you want semantic analysis to do here is not to mark when a block contains a "label". What you want to check is whether the block can be jumped into; whether there is control flow that can enter the block at another position than the block's start. I suggest changing the naming of a number of variables according to that. Something like `hasMultipleEntryPoints`.

Indeed it's not the best name. The problem is, I don't see a way of doing it with just one variable.
Except if you refer just to the naming, meaning just changing the names to something like:
```
hasLabel = ...
hasCaseLabel = ...
hasMultipleEntryPoints = hasLabel || hasCaseLabel;
if (hasMultipleEntryPoints) { ...
```

thewilsonator commented:
This good to go?

baziotis commented:
> This good to go?

If you don't mind, I'd like to wait for @JohanEngelen 's answer and then it will be ready. Thanks!

WalterBright commented:
> Because it helps LDC back-end do dead-code elimination when the condition is constant

Dead code elimination is normally done by the optimizer. I'm baffled why the LLVM optimizer doesn't do this already? (The much-maligned DMD optimizer does! Gotos and labels don't mess it up.)

WalterBright commented:
> But if the body to-be-elided has any label

Labels are rare. But you can test funcdecl.labtab to see if there are any, and so skip processing if there are. You can also set a flag if there are any switch statements, again skipping the processing if there aren't any. I think this would be a better solution than adding memory for each statement.

But I still don't understand why the LLVM optimizer can't handle gotos.

baziotis commented:
> > Because it helps LDC back-end do dead-code elimination when the condition is constant
> 
> Dead code elimination is normally done by the optimizer. I'm baffled why the LLVM optimizer doesn't do this already? (The much-maligned DMD optimizer does! Gotos and labels don't mess it up.)

LLVM does it. With this, the reasoning is that we just make its life easier. Or that we make the code
better if the optimizer is not invoked at all (`-O0`). Because the idea is to do it before
we call LLVM as Clang does.

 > Labels are rare. But you can test funcdecl.labtab to see if there are any, and so skip processing if there are.

Sorry, I'm not in front of a computer right now. But I assume that `funcdecl.labtab` tells us if there are any labels at all. What I mean is that what we want is to know if there are inside any specific scopes
inside the function (the scopes of conditional statements).

> You can also set a flag if there are any switch statements, again skipping the processing if there aren't any.

Again, this I think does not achieve the desired effect because a switch might be in a scope
that we don't care, or it might not have any case that acts as a label.

One thing to consider of course is the rarity of both of those. But I think it's simple to do anyway.

> I think this would be a better solution than adding memory for each statement.

This is true. I wanted to make it so that only `ScopeStatement` had such info but did not
find the time.

baziotis committed:
Set info only in ScopeStatement and cleanup

baziotis commented:
This should be better. It sets only `ScopeStatement` and I cleaned up the naming a bit.
Tell me your opinion. Btw, this will have more tests, they're not enough.
One thing I don't like is that do this for all cases, we have to pass the scope info up in _every_
point that a `ScopeStatement` can exist. Is there a way when we pop a scope to copy the `hasLabel`
and `hasCase` fields ?

baziotis commented:
Anyone interested in this ?
